reasoner_pydantic/base_model.py

```diff
@@ -0,0 +1,48 @@
+from typing import Callable
+import weakref
+from pydantic import BaseModel as PydanticBaseModel, PrivateAttr
+
+class BaseModel(PydanticBaseModel):
+    """
+    Custom base model for all classes
+
+    This provides a hash function that assumes all fields are either:
+
+    1. Immutable
+    2. Derived from BaseModel
+    3. Able to call a hash invalidation hook on their own
+    """
+
+    _hash: int = PrivateAttr(default=None)
+    _invalidate_hook: Callable = PrivateAttr(default=None)
+
+    def __hash__(self) -> int:
+        """ Hash function based on Pydantic implementation """
+        if not self._hash:
+            self._hash = hash(self.__class__) + hash(tuple(self.__dict__.values()))
```

Comment:
Is there a reason for summing the hashes instead of 
```
self._hash = hash((self.__class__, tuple(self.__dict__.values())))
```
as the docs recommend: https://docs.python.org/3/reference/datamodel.html#object.__hash__

Reply:
Didn't realize that was the recommended behavior, I was going off of the Pydantic implementation. I'll update it.